Decouplage des données de Github et GoogleSheet

eonm-pro opened the pull request:
Cette PR propose de découpler davantage les données de GitHub et de GoogleSheet

**Données au format RDF (profil d'application et ontologie)**

* Sur le plan technique : les données RDF sont générées à partir de GoogleSheet et versionnées dans Github grâce à une action GitHub (`google-sheet-to-ttl` qui est déclenchée manuellement. **Le Dockerfile utilise les données du dépôt pour générer la documentation de l'ontologie et du profil d'application**, ce qui fiabilise le processus de build et permet de redéployer des versions antérieures.
* Dans le dépôt il n'y a plus de distinction entre  les données profil d'application et celles l'ontologie. **L'ensemble des données de rdafr sont réunies dans le dossier ontologie**. Les fichiers du profil d'application et de l'ontologie étant identiques je ne voyais pas l'intérêt de dupliquer deux fois les mêmes fichiers dans deux dossiers différents (`ontologie `et `profil-application`). La distinction entre profil d'application et ontologie s’opère au moment de la génération des pages HTML avec Widoco et ShaclPlay
* Les métadonnées de l'ontologie sont générées grâce au fichier `ontologie/config.properties`.

**Dockerfile**

* J'ai opté pour une version slim de debian pour la partie build du Dockerfile

eonm-pro committed:
Fix: utilisation de debian slim

eonm-pro committed:
Fix: changement du répertoire d'installation de Widoco

évite que Widoco ne se retrouve sur l'image docker en production

eonm-pro committed:
Feat: Sauvegarde de données de Google Sheet

L'action transforme les données de l'ontologie dans différents formats RDF et les commit dans le dépôt.

eonm-pro committed:
Feat: Ajout des données rdf de l'ontologie

eonm-pro committed:
Feat: Ajout des métadonnées de l'ontologie

Les métadonnées sont lues par Widoco lors de la génération de la documentation de l'ontologie

eonm-pro committed:
Feat: Ajout des métadonnées pour l'ontologie

eonm-pro committed:
Fix: Suppression du dossier profil d'application

Les données RDF de l'ontologie et du profil d'application (qui sont les mêmes fichiers) sont stocké dans le dossier ontologie

eonm-pro committed:
Feat: Optimisation du Dockerfile

Les maj apt sont dans le premier layer

eonm-pro committed:
Fix: métadonnées de l'ontologie

kerphi reviewed (commented):
Je me pose une question sur le fichier google docs : 
https://docs.google.com/spreadsheets/d/1CZIf3bxuuH3aghFn7B7P89DrLqp_jzv4moI_BpI9PzY

Et le nouveau fichier de configuration : 
ontologie/config.properties 

Ne risque-t-on pas d'oublier de mettre à jour ontologie/config.properties quand on modifier le google docs ? Je serais plus serein si ce fichier ontologie/config.properties été généré à partir du google-docs car son contenu est qd même très lié au contenu du google docs non ? si on place les deux fichiers (gdocs et config.properties) à deux endroits différents on éparpille l'info et on risque d'avoir des loupés qd les mises à jours auront lieux.

Ne pourrait on pas dédier un onglet dans le google docs pour y mettre toute cette configuration ? A la rigueur un simple onglet à deux colonnes : colonne 1 pour les clés et colonne 2 pour les valeurs ? ou autre.

kerphi commented on Dockerfile:
```diff
@@ -12,6 +8,11 @@ COPY ./siteweb/.docker/*   /build/
+RUN mkdir /build/
+COPY ./siteweb/*   /build/
+COPY ./siteweb/.docker/*   /build/
+COPY ./ontologie /ontologie
```
Je note ici une erreur que je ne comprends pas dans le [build sur la github action](https://github.com/transition-bibliographique/ontologie-rda-fr/actions/runs/4977716455/jobs/8910371021#step:3:1077) : 

```
Step 6/28 : COPY ./ontologie /ontologie
failed to export image: failed to create image: failed to get layer sha256:a63b7c35f37d5ad72b54658b662c9638f7959a1cf36aa3a06783aeece4e53bb4: layer does not exist
Error: Process completed with exit code 1.
```

kerphi replied:
A noter : j'ai tenté de builder en local et je n'ai pas l'erreur.

eonm-pro replied:
> Je note ici une erreur que je ne comprends pas dans le [build sur la github action](https://github.com/transition-bibliographique/ontologie-rda-fr/actions/runs/4977716455/jobs/8910371021#step:3:1077) :

Salut @kerphi Je n'ai pas d'erreur non plus en local. Je ne comprends pas non plus d'où vient l'erreur. Peut être une histoire de branche ? Ontologie n'existe pas sur main.

> Ne risque-t-on pas d'oublier de mettre à jour ontologie/config.properties quand on modifier le google docs ? Je serais plus serein si ce fichier ontologie/config.properties été généré à partir du google-docs car son contenu est qd même très lié au contenu du google docs non ? si on place les deux fichiers (gdocs et config.properties) à deux endroits différents on éparpille l'info et on risque d'avoir des loupés qd les mises à jours auront lieux.

Pour ce point, on a les métadonnées du google sheet qui servent à générer les méta du profil d'application et le fichier `config.properties` les métadonnées de l'ontologie. Aujourd'hui les métadonnées de l'ontologie et du profil sont très similaires, mais est-ce que ça sera le cas tout le temps ?

Si on récupère les données depuis Google Sheet,  on devra patcher au cas par cas les métadonnées qui divergent (pour l'abstract par exemple), est-ce préférable ? Où stockons nous les données qu'il faut patcher ?

eonm-pro replied:
> Ne pourrait on pas dédier un onglet dans le google docs pour y mettre toute cette configuration ? A la rigueur un simple onglet à deux colonnes : colonne 1 pour les clés et colonne 2 pour les valeurs ? ou autre.

Si on décide de mutualiser les métadonnées dans le fichier google. Il faudrait un onglet pour le profil et un onglet pour l'ontologie. Quand on génère l'un ou l'autre il faut réussir à enlever les métadonnées qui ne conviennent pas, ce qui implique de manipuler le fichier excel.

Ou sinon on fait deux fichiers de métadonnées qu'on stocke sur Github. On exporte les métadonnées en RDF. Avant de lancer Widoco et ShaclPlay  on concatène le jeu de métadonnées qui vont bien au fichier.

Là ou ça risque de coincer c'est avec l'outil https://xls2rdf.sparna.fr/rest/, je ne sais pas si on peut envoyer un fichier à l'API. La documentation de l'API ne dit rien, mais ça semble possible avec l'interface : https://skos-play.sparna.fr/play/convert.

kerphi commented on Dockerfile:
```diff
@@ -12,6 +8,11 @@ COPY ./siteweb/.docker/*   /build/
+RUN mkdir /build/
+COPY ./siteweb/*   /build/
+COPY ./siteweb/.docker/*   /build/
+COPY ./ontologie /ontologie
```
Ca ne serait pas plutôt ça comme commande ?
```
COPY ./ontologie/ /build/ontologie/
```

eonm-pro replied:
J'ai copié les données RDF dans le dossier /ontologie pour éviter de polluer le dossier /build/ontologie avec des données parasites qui ne servent que pour la génération de la documentation.

Peut être que mettre les données rdf dans /tmp/ontologie serait plus clair ? Ou sinon on laisse les données "parasites", mais elles seront servies par ngnix si la personne connait le nom du fichier.

kerphi commented on Dockerfile:
```diff
@@ -46,44 +47,31 @@ RUN pandoc --standalone \
       -noPlaceHolderText \
       -ignoreIndividuals
 
+RUN mv ./doc ./ontologie
```
d'où vient ``./doc`` ? c'est un fichier, un répertoire, pas quel script il est généré ?

eonm-pro replied:
Le dossier doc est auto généré par Widoco lorsqu'on utilise le flag -confFile. Ce n'était pas le cas lorsqu'on récupérait les métadonnées directement depuis le fichier nt. Dans ce cas il mettait l'ensemble des fichiers à la racine du dossier spécifier avec le flag -outFolder. Avec -configFile il créé un dossier doc dans le dossier spécifié par -outFolder.

Il y a une petite inconsistance on dirait.

Je génère la documentation dans le dossier courant (qui se retrouve avec ./doc) puis je copie le contenu de doc dans le dossier ontologie pour avoir la structuration qu'on a actuellement sur le site web.

kerphi commented on Dockerfile:
```diff
@@ -46,44 +47,31 @@ RUN pandoc --standalone \
-
-RUN sed -e "#http://www.w3.org/ns/shacl#d" -e "/_:node/d" -i ./rdafr.nt
+# Enlève les noeuds vides et les éléments shacl qui sont problématiques pour Widoco
+RUN sed -e "#http://www.w3.org/ns/shacl#d" -e "/_:node/d" -i /ontologie/rdafr.nt
```
Attention je pense qu'on va patcher ici avec le sed dans le répertoire /ontologie/ alors qu'on devrait plutôt aller patcher dans /build/ontologie/ si je suis la logique mis en place.

Je pense que c'est la ligne ``WORKDIR /build/`` qui prête à confusion car une fois qu'on a placé ce WORKDIR on va mettre plein de chemins relatifs avec ``./`` ce qui à mon sens brouille un peu la lisibilité du script. Pourquoi ne pas utiliser de partout des chemins absolus ?
Et à chaque fois indiquer explicitement : 
```
/build/ontologie/rdafr.nt
/build/ontologie/config.properties
/build/ontologie/index.html
[...]
```

eonm-pro replied:
> Attention je pense qu'on va patcher ici avec le sed dans le répertoire /ontologie/ alors qu'on devrait plutôt aller patcher dans /build/ontologie/ si je suis la logique mis en place.

voir supra : https://github.com/transition-bibliographique/ontologie-rda-fr/pull/9#discussion_r1193739937

> Pourquoi ne pas utiliser de partout des chemins absolus ?

Ok, je fais ça une fois la question https://github.com/transition-bibliographique/ontologie-rda-fr/pull/9#discussion_r1193739937 résolue

eonm-pro replied:
J'ai fait les modifs. Sauf pour certains chemins de pandoc. Cela provoquait des bug de génération.

eonm-pro committed:
Fix: Emplacement de l'ontologie

Simplification des chemins
Ajout de documentation

eonm-pro committed:
Fix: typo

eonm-pro committed:
Feat: Ajout des fichiers de métadonnées pour le profil et l'ontologie

eonm-pro committed:
Feat: Optimisation Dockerfile

L'installation de Widoco est réalisée en début de build, pour éviter de reconstruire le layer quand seules les données changent

eonm-pro committed:
Fix: réécriture des chemins

eonm-pro committed:
Doc: Ajout de commentaires

eonm-pro committed:
Fix: Mise à disposition du profil d'application au format ttl

Fix un lien cassé sur  le site rdafr.fr

eonm-pro committed:
Fix: problème de build du Dockerfile sur GitHub

eonm-pro committed:
Fix: chemins absolu

Utilisation de chemins absolu sauf pour certaines paramètre de pandoc (provoque des bug pour le style)

eonm-pro commented:
@kerphi j'ai fait les modifs demandées. Pour les métadonnées ça fonctionne. J'ai testé en local.

En revanche on a toujours l'erreur lors du build sur GittHub. Un problème de cache ?

eonm-pro committed:
Fix: fix problème de build ave Docker

Avec un dossier de destination c'est mieux !
